promotion/src/main/java/org/cryptimeleon/incentive/promotion/PromotionDescription.java

```diff
@@ -0,0 +1,54 @@
+package org.cryptimeleon.incentive.promotion;
+
+
+import lombok.Getter;
+
+import java.time.LocalDate;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Abstract class that represents a definition of a promotion.
+ */
+@Getter
+abstract public class PromotionDescription {
+    long promotionId;
+    String promotionTitle;
+    String promotionDescription;
+    LocalDate promotionStart;
+    LocalDate promotionEnd;
+    List<PromotionReward> promotionRewards;
+
+    public PromotionDescription(long promotionId,
+                                String promotionTitle,
+                                String promotionDescription,
+                                LocalDate promotionStart,
+                                LocalDate promotionEnd,
+                                List<PromotionReward> promotionRewards) {
+        this.promotionId = promotionId;
+        this.promotionTitle = promotionTitle;
+        this.promotionDescription = promotionDescription;
+        this.promotionStart = promotionStart;
+        this.promotionEnd = promotionEnd;
+        this.promotionRewards = promotionRewards;
+    }
+
+    /**
+     * Verify that this promotion is valid on the given date.
+     */
+    public boolean isValidAt(LocalDate date) {
+        return !date.isBefore(promotionStart) && !date.isAfter(promotionEnd);
+    }
+
+    /**
+     * Computes the points that are can be earned for this promotion.
+     */
+    public abstract Long computePoints(Basket basket);
+
+    /**
+     * Return a list of all rewards for which the given token points count qualifies one.
+     */
+    public List<PromotionReward> qualifiedRewards(long points) {
+        return promotionRewards.stream().filter(promotionReward -> promotionReward.getPrice() <= points).collect(Collectors.toList());
```

Comment:
More generally, I feel like this method should get as input:
- the current basket content
- a description of the token (or what the provider learns about the token content?! Not sure how this can be made cleverly symmetric)

I haven't thought this through completely. But I feel like that structure would support things like "I need to buy two hazelnut spread thingies to get one for free" better, as it depends on basket content whether you're buying enough and on token content, how many you've already bought before.

Reply:
Good point, I will add that!